addons/web/controllers/main.py

```diff
@@ -1575,6 +1575,10 @@ def report_routes(self, reportname, docids=None, converter=None, **data):
             pdf = report.with_context(context).render_qweb_pdf(docids, data=data)[0]
             pdfhttpheaders = [('Content-Type', 'application/pdf'), ('Content-Length', len(pdf))]
             return request.make_response(pdf, headers=pdfhttpheaders)
+        elif converter == 'text':
```

Comment:
[ADD] base: report: qweb-text type

This commit adds a new ir.action.report type: qweb-text. It produces text files, serves them with a text mimetype and these reports use the same rendering context as the current other types.

This will be used in a following commit to produce ZPL scripts.

Reply:
@sle-odoo I'm working on ESC-POS / EPL / FPL printing. It would be nice if this controller method could be made more easy to inherit. It is uneasy to extend as `docids` and `data` are modified in place. Would be nice, if this feedback could be heared.

Reply:
Note: I could standardize on `qweb-text` but at least for the esc-pos I'd like to use a preprocessor, so here the use case holds true.

Reply:
hi @blaggacao,
what do you suggest to improve the inheritance of this controller? do you want to be able to add a new type to ir.actions.report and not having to worry about the `report_download` and `report_routes` methods? i thought about that but refactoring these methods would have made the diff too big and we'd need to guess or store somewhere the mimetypes to serve these new types.

Reply:
Quite right. But never mind, I found a better way... :)
I back ported the python side of this commit to v11 and added a field `raw_format` on `ir.actions.report` with a selection as per the mimetype (`zpl`, `epl`, `fpl`, `escpos`). Then, I override the `render_qweb_text` and add post-processors based on `self.raw_format`. As the client get's the complete report action's `read()` (at least in v10 it was so), also the client will be aware of the mimetype whie handinlg the report action.

Therefore, `text` becomes an excellent raw format transport.

:+1: Nice commit, reduces my code roughly about 150 lines :)

Reply:
```python

class IrActionsReportXml(models.Model):
    _inherit = 'ir.actions.report'

    direct_printing = fields.Boolean()
    automatic_printing = fields.Boolean()
    raw_format = fields.Selection([
        ('zpl, "ZPL'),
        ('fpl', "FPL"),
        ('epl', "EPL"),
        ('escpos', "ESC/POS")
    ], "Raw text format")


    @api.onchange('report_type')
    def _onchange_report_type(self):
        if self.report_type not in ['qweb-pdf',
                                    'qweb-text']:
            self.direct_printing = self.automatic_printing = False
        if self.report_type not in ['qweb-text']:
            self.raw_format = False

    def _post_escpos(self, data):
        # FIXME: Add postprocessor here
        return data

    @api.model
    def render_qweb_text(self, docids, data=None):
        data = self._get_rendering_context(docids, data)
        res = self.render_template(self.report_name, data), 'text'
        if self.raw_format == 'escpos':
            res = self._post_escpos(res)
        return res

```

Reply:
👍

Reply:
```python
from escpos import printer, escpos
from xmlescpos import Layout
...
    def _post_escpos(self, data):
        with escpos.EscposIO(printer.Dummy()) as prntr:
            Layout(data).format(prntr.printer)
            res = prntr.printer.output
        return res
```
with (originally) Frederic's https://github.com/xoe-labs/py-xml-escpos